torch/optim/__init__.py

```diff
@@ -6,21 +6,21 @@
 future.
 """
 
-from . import lr_scheduler, swa_utils
```

Comment:
Is ufmt forcing the removal of all these relative imports?

Reply:
No. This is updated by hand.

> #### Absolute imports, or relative imports from siblings, are recommended by [PEP8](https://www.python.org/dev/peps/pep-0008/).

> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured (such as when a directory inside a package ends up on `sys.path`):
> 
> ```python
> import mypkg.sibling
> from mypkg import sibling
> from mypkg.sibling import example
> ```
> 
> However, explicit relative imports are an acceptable alternative to absolute imports, especially when dealing with complex package layouts where using absolute imports would be unnecessarily verbose:
> 
> ```python
> from . import sibling
> from .sibling import example
> Standard library code should avoid complex package layouts and always use absolute imports.
> ```

------

We can enforce this by adding the [ruff rule `TID` (`flake8-tidy-imports`)](https://docs.astral.sh/ruff/rules/#flake8-tidy-imports-tid) with:

```toml
[tool.ruff.lint.flake8-tidy-imports]
ban-relative-imports = "all"  # defaults to "parents"; see https://docs.astral.sh/ruff/settings/#lintflake8-tidy-imports
```

Reply:
Can't this affect import ordering which can be bad for our Python submodules in torch that do have side effects (or cause cyclic import errors where non existed before).

Reply:
Changing relative import to identical absolute import does not change the side effects of the import statement.

Reply:
Doesn't it change the import order though?
Where the relative import would actually trigger the import of the file while the absolute import would expect the partially initialized module to already contain the corresponding entry?

Reply:
This change is made based on:

1. The relative import will convert to absolute import in bytecode eval loop:

```console
$ echo 'from torch.optim import adam' | python3 -m dis -
  0           0 RESUME                   0

  1           2 LOAD_CONST               0 (0)
              4 LOAD_CONST               1 (('adam',))
              6 IMPORT_NAME              0 (torch.optim)
              8 IMPORT_FROM              1 (adam)
             10 STORE_NAME               1 (adam)
             12 POP_TOP
             14 RETURN_CONST             2 (None)
$ echo 'from . import adam' | python3 -m dis -
  0           0 RESUME                   0

  1           2 LOAD_CONST               0 (1)
              4 LOAD_CONST               1 (('adam',))
              6 IMPORT_NAME              0
              8 IMPORT_FROM              1 (adam)
             10 STORE_NAME               1 (adam)
             12 POP_TOP
             14 RETURN_CONST             2 (None)
```

https://github.com/python/cpython/blob/a9f2daf1ab182d95b44ee94dc9fb8faec60e34b1/Python/ceval.c#L2513-L2526

`from . import adam` is convert to `__import__(f'{__name__}.adam')` where `__name__ = 'torch.optim'`.
	
They are semantically identical.

2. See https://github.com/pytorch/pytorch/pull/127703#discussion_r1623617154: Absolute imports, or relative imports from siblings, are recommended by [PEP8](https://www.python.org/dev/peps/pep-0008/).

Reply:
How about absolute paths when combined with circular dependencies? Example:

torch/optim/__init__.py import `torch.optim._multi_tensor` and
torch/optim/_multi_tenoser/__init__.py import `torch.optim`

This pattern, combined with absolute import paths is causing unpredictable and hard to debug failures.